File: single_layer_perceptron/Matlab/Perceptron.m
classdef Perceptron < handle
    % Perceptron class
    
    properties
        n_features
        weight
        learning_rate
        n_iteration
        dw
    end
    
    methods
        function obj = Perceptron(n_features, learning_rate, n_iteration)
            % Initialize the perceptron
            obj.n_features = n_features;
            obj.weight = randn(n_features + 1, 1);
            obj.learning_rate = learning_rate;
            obj.n_iteration = n_iteration;
            obj.dw = 0;
        end
        
        function output = sigmoid(obj, x)
            % Return the sigmoid of x
            output = (2 ./ (1 + exp(-x))) - 1;
        end
        
        function result = forward(obj, x)
            % Return the output of the perceptron with input x and weight w
            result = obj.sigmoid(obj.weight' * x);
        end
        
        function grad_result = grad(obj, x, y)
            % Compute the gradient of the weight
            output = obj.forward(x);
            obj.dw = sum(-0.5 .* (y - output) .* (1 - output.^2) .* x, 2);
            grad_result = obj.dw;
        end
        
        function updated_weight = step(obj)
            % Update the weight of the perceptron with the gradient
            obj.weight = obj.weight - obj.learning_rate * obj.dw;
            updated_weight = obj.weight;
        end
        
        function train(obj, x, y)
            % Train the perceptron
            for i = 1:obj.n_iteration
                obj.grad(x, y);
                obj.step();
            end
        end
        
        function prediction = predict(obj, x)
            % Return the prediction of the perceptron
            prediction = obj.forward(x);
        end
    end
end

File: single_layer_perceptron/perceptron.py
"""Define the perceptron class"""
import numpy as np


class Perceptron:
    """Perceptron class."""

    def __init__(self, n_features, learning_rate=0.01, n_iteration=1000):
        """
        Initialize the perceptron.

        Parameters:
        ----------
            n_features: int
                Number of features of the input data
            learning_rate: float
                Learning rate of the perceptron
            n_iteration: int
                number of iteration to train the perceptron
            weight: numpy array
                Weight of the perceptron with shape (n_features + 1, 1)
                it includes the bias term
            dw: numpy array
                The gradient of the weight
        ----------
        """
        self.n_features = n_features
        self.weight = np.random.normal(size=(n_features + 1, 1))
        self.learning_rate = learning_rate
        self.n_iteration = n_iteration
        self.dw = 0

    def sigmoid(self, x):
        """Return the sigmoid of the x."""
        return (2 / (1 + np.exp(-x))) - 1

    def forward(self, x):
        """Return the output of the perceptron with input x and weight w."""
        return self.sigmoid(np.dot(self.weight.T, x))

    def grad(self, x, y):
        """
        Compute the gradient of the weight.

        Parameters:
        ----------
            x: numpy array
                input of the perceptron with shape (n_features + 1, 1)
                it includes the bias term
            y: numpy array
                target of the perceptron with shape (1, 1)

        """
        output = self.forward(x)
        self.dw = np.sum(
            -0.5 * (y - output) * (1 - output**2) * x, 1, keepdims=2
        )
        return self.dw

    def step(self):
        """Update the weight of the perceptron with the gradient."""
        self.weight = self.weight - self.learning_rate * self.dw
        return self.weight

    def train(self, x, y):
        """Train the perceptron."""
        for _ in range(self.n_iteration):
            self.grad(x, y)
            self.step()

    def predict(self, x):
        """Return the prediction of the perceptron."""
        return self.forward(x)
